### DISCUSSION TITLE
Any plans to implement Apollo Gateway?

### DISCUSSION THREAD

**Discussion by command-tab** [mirumee/ariadne#496] [Ideas]
Looking at the [Federated Gateway docs](https://ariadnegraphql.org/docs/apollo-federation#federated-gateway), I see the example uses Apollo Gateway to federate related Ariadne GraphQL services. Are there any plans to replicate Apollo Gateway such that a complete federated graph can be run on _only_ Python, rather than mostly Python and some Javascript? Thanks!

**Comment by rafalp** [comment]
Hi, thanks for asking!

We've done a quick look on implementing federation composition capabilities but decided against it (for now). It would take considerable amount of time that we could've placed elsewhere + there's usually no problem in proxying your Ariadne services behind Apollo-Server in production environments.

**Comment by Ambro17** [comment]
+1 To implementing the first intelligent gateway in python. I'm open to contribute

**Comment by sebastiandev** [comment]
So I might be missing something, but, in our case we have a custom throttling logic as well as authentication. Where would that go? I was thinking that in the gateway since its the entry point to the rest of the services, otherwise I would have to replicate (re-use) on every service? Which makes no sense because throttling and auth have to run before any query/mutation is executed.

Its not very clear how and where to implement those sort of things and the gateway js part makes it even more confusing.

**Reply by rafalp** [reply]
Auth: apollo-server should proxy credentials to downstream services

Throttling: apollo-server

**Reply by sebastiandev** [reply]
which means we have to port our python code to javascript.. meaning we can't make use of ariadne for that... this js + python is a bit cumbersome, I don't think this fits well with our plan which is sad because I really wanted to give this a try
